src/System.Private.CoreLib/shared/System/Diagnostics/Tracing/CounterGroup.cs

```diff
@@ -205,48 +244,66 @@ private void OnTimer(object? state)
-        {
-            private int _period;
-            private TimerCallback _callback;
-            private object _state;
+        private static Thread? s_pollingThread;
+        private static ManualResetEvent? s_pollingThreadEvent;
+        // Used for sleeping for a certain amount of time while allowing the thread to be woken up
+        private static AutoResetEvent? s_pollingThreadSleepEvent;
 
-            internal Timer(TimerCallback callback, object state, int dueTime, int period)
-            {
-                _callback = callback;
-                _state = state;
-                _period = period;
-                Schedule(dueTime);
-            }
+        private static List<CounterGroup>? s_counterGroupEnabledList;
 
-            private void Schedule(int dueTime)
+        private static void PollForValues()
+        {
+            ManualResetEvent? pollingThreadEvent = null;
+            AutoResetEvent? sleepEvent = null;
+            while (true)
             {
-                Task.Delay(dueTime, Token).ContinueWith(OnTimer, null, CancellationToken.None, TaskContinuationOptions.ExecuteSynchronously | TaskContinuationOptions.OnlyOnRanToCompletion, TaskScheduler.Default);
-            }
+                int sleepDurationInMilliseconds = Int32.MaxValue;
+                lock (s_counterGroupLock)
+                {
+                    pollingThreadEvent = s_pollingThreadEvent;
+                    sleepEvent = s_pollingThreadSleepEvent;
 
-            private void OnTimer(Task t, object s)
-            {
-                Schedule(_period);
-                _callback(_state);
+                    if (s_counterGroupEnabledList != null)
+                    {
+                        foreach (CounterGroup counterGroup in s_counterGroupEnabledList)
+                        {
+                            DateTime now = DateTime.UtcNow;
+                            if (counterGroup._nextPollingTimeStamp < now + new TimeSpan(0, 0, 0, 0, 1))
+                            {
+                                counterGroup.OnTimer();
+                            }
+
+                            int millisecondsTillNextPoll = (int)((counterGroup._nextPollingTimeStamp - now).TotalMilliseconds);
+                            millisecondsTillNextPoll = Math.Max(1, millisecondsTillNextPoll);
+                            if (millisecondsTillNextPoll < sleepDurationInMilliseconds)
+                            {
+                                sleepDurationInMilliseconds = millisecondsTillNextPoll;
+                            }
+                        }
+                    }
+                }
+                // Only sleep if we actually had something in the enabled counter group
+                if (sleepDurationInMilliseconds != Int32.MaxValue)
+                {
+                    sleepEvent?.WaitOne(sleepDurationInMilliseconds);
+                }
+                if (pollingThreadEvent != null)
```

Comment:
With the addition of the AutoResetEvent we don't also need a ManualResetEvent (as far as I can tell the behavior is correct, just more complicated than necessary). I threw together a little PR that simplified this, as well as some other stuff. Don't feel compelled to take it, I just find it a clearer way to make a suggestion than text in some cases. https://github.com/sywhang/coreclr/pull/1